### DISCUSSION TITLE
What has been your experience using NextJS + Tauri?

### DISCUSSION THREAD

**Discussion by serg06** [tauri-apps/tauri#6083] [Q&A] [answered]
Hello! I'm looking to start a new Tauri project. I'm wondering whether I should use NextJS or React. I've used both for personal projects, but I notice that NextJS runs a lot slower in development mode.

How has your experience been with Tauri + NextJS? Does it take a long time to start up? Does HMR work well?

Would love to hear your thoughts!

**Comment by FabianLars** [comment, marked as answer]
> Does it take a long time to start up?

Yes, just like in a standalone nextjs app, it's just a heavy dev environment, and it uses webpack (i'm not taking turbopack into account yet) which is also pretty slow.

> Does HMR work well?

Yes, Tauri isn't any different from any normal browser in that regard.

The main problem imo with nextjs in a tauri app, is its SSR-first nature. For nextjs to work in tauri you need to use its SSG mode (`next export`) which can get a bit annoying at times.

Personally i just go with vite + react, but i also prefer react-router (or similar) over file-based routers like the one in nextjs, so take that with a grain of salt i guess.

**Reply by smolattack** [reply]
There's also TanStack router which is a lot more TypeScript friendly. I think a big advantage of nextjs is the automatic lazyfication of routes which has zero value in a desktop application. A 100mb page is terrible but a desktop app occupying 10x that much space in RAM is meh (most of your users won't notice it and even if they did, probably wouldn't care). So the problems nextjs is trying to solve aren't necessarily an advantage for tauri apps.

**Reply by Manubi** [reply]
Is anyone building a nextjs app and a tauri app with nextjs and can tell something about the devflow and dx?

**Reply by jhenderson-moshipco** [reply]
I'm running into some of these issues right now. The problem with having Next.js running in Tauri is that next already has the server side environment built in, unlike plain React or Vite. 

Our application frontend lives in a monolithic repo where we've created our own workspace packages to handle UI, data fetching, types, utils, context, etc and we have a mobile, desktop, and web application folder that's just a thin app layer that we inject all the workspace packages into. It made sense to do this cause we'd have CI/CD, tests, and all of our utilities written in one place. 

We didn't want to move off of next because we originally just had the web app built and it was in Next.js. To make all the components and utility functions we have work, we needed to stay in the next environment. The thing is, Next.js isn't really meant for this. It's best suited as a web environment as @smolattack said (pre-fetching, optimized lazy loads, SSR, etc.). 

If you're doing anything wit Next.js [slug] pages, you'll run into Next.js build errors & the Tauri build will get stuck if Next.js is not configured properly. It's kind of a headache, but we've found solutions. I'd say if you can get around not using Next.js with Tauri that's what I would do. It's kinda like trying to put two different engines in a car and hoping they work together.

**Reply by daveycodez** [reply]
TanStack Start is much better suited for Tauri. And most other things at this point 👍🏼

**Comment by danieltan007** [comment]
still no info about using ssr?

**Reply by FabianLars** [reply]
Like in my other comment, SSR doesn't work in Tauri because there is no nodejs runtime, nothing will change about that anytime soon. So if nextjs + SSR is a _must_ you may be able to bundle your nextjs app via [pkg](https://github.com/vercel/pkg) as a [sidecar](https://tauri.app/v1/guides/building/sidecar/).

**Reply by Dimanchecfa** [reply]
> Toujours aucune information concernant l'utilisation du SSR ?

  Next.js avec Tauri exige le mode SSG (output: 'export'). Ça veut dire :
  - Pas de Server Components                                                                                                                                                                                                       
  - Pas de server actions                                                                                                                                                                                                          
  - Pas de next/image optimisé                                                                                                                                                                                                     
  - Pas d'API routes
  - Pas de middleware Next.js

 En gros, tu perds tout ce qui fait l'intérêt de Next.js. Il te reste un React pur avec un file-based router — autant utiliser Vite directement.

**Comment by daveycodez** [comment]
TanStack Start is superior for Tauri / Capacitor, you just turn on SPA mode or pre-rendering and use the client loaders / React Query for data fetching
